proofs/proof.ml

```diff
@@ -544,6 +544,21 @@ let refine_by_tactic ~name ~poly env sigma ty tac =
   let (ans, _) = Safe_typing.inline_private_constants env ((ans, Univ.ContextSet.empty), neff) in
   ans, sigma
 
+(* Prevent side effects from leaking out, cf #13324
+   It would be nicer if side effects were contained in the evar map
+   (also then we may be able to avoid inlining them here)
+   but that requires more extensive change.
```

Comment:
Are we able to evaluate the amount of work it would require?

Reply:
This is part of our informal roadmap, but it basically amounts to a redesign of abstract and it's death by a thousand of paper cuts.

Reply:
Sorry, pas compris. Why only `abstract`? And who is "we" in "our"?

My English level is probably also too low. Should "it's death by a thousand of paper cuts" be understood as, maybe, "it will require looking closely at many related places of the code"?

Reply:
It is a wild guess, but maybe if we could keep pushing `Global.env ()` upwards and maybe after pretying, we could try to understand what kind of dances we are doing with the end and try to manage them.